compiler/src/dotty/tools/backend/jvm/GenBCode.scala

```diff
@@ -48,34 +22,67 @@ class GenBCode extends Phase {
     superCallsMap.update(sym, old + calls)
   }
 
-  private val entryPoints = new mutable.HashSet[String]()
-  def registerEntryPoint(s: String): Unit = entryPoints += s
+  private[jvm] val entryPoints = new mutable.HashSet[String]()
+  def registerEntryPoint(s: String): Unit =
+    entryPoints += s
 
-  private var myOutput: AbstractFile = _
+  private var _backendInterface: DottyBackendInterface = _
+  def backendInterface(using Context): DottyBackendInterface = {
+    if _backendInterface eq null then
+      _backendInterface = DottyBackendInterface(superCallsMap)
+    _backendInterface
+  }
 
-  private def outputDir(using Context): AbstractFile = {
-    if (myOutput eq null)
-      myOutput = ctx.settings.outputDir.value
-    myOutput
+  private var _codeGen: CodeGen = _
+  def codeGen(using Context): CodeGen = {
+    if _codeGen eq null then
+      val int = backendInterface
+      val dottyPrimitives = new DottyPrimitives(ctx)
+      _codeGen = new CodeGen(int, dottyPrimitives)(bTypes.asInstanceOf[BTypesFromSymbols[int.type]])
+    _codeGen
   }
 
-  private var myPrimitives: DottyPrimitives = null
+  private var _bTypes: BTypesFromSymbols[DottyBackendInterface] = _
+  def bTypes(using Context): BTypesFromSymbols[DottyBackendInterface] = {
+    if _bTypes eq null then
+      _bTypes = BTypesFromSymbols(backendInterface, frontendAccess)
+    _bTypes
+  }
 
-  override def run(using Context): Unit =
-    if myPrimitives == null then myPrimitives = new DottyPrimitives(ctx)
-    new GenBCodePipeline(
-      DottyBackendInterface(outputDir, superCallsMap),
-      myPrimitives
-    ).run(ctx.compilationUnit.tpdTree)
+  private var _frontendAccess: PostProcessorFrontendAccess = _
+  def frontendAccess(using Context): PostProcessorFrontendAccess = {
+    if _frontendAccess eq null then
+      _frontendAccess = PostProcessorFrontendAccess.Impl(backendInterface)
+    _frontendAccess
+  }
 
+  private var _postProcessor: PostProcessor = _
+  def postProcessor(using Context): PostProcessor = {
+    if _postProcessor eq null then
+      _postProcessor = new PostProcessor(frontendAccess, bTypes)
+    _postProcessor
+  }
 
-  override def runOn(units: List[CompilationUnit])(using Context): List[CompilationUnit] = {
-    outputDir match
-      case jar: JarArchive =>
-        updateJarManifestWithMainClass(jar, entryPoints.toList)
-      case _ =>
-    try super.runOn(units)
-    finally outputDir match {
+  override def run(using Context): Unit =
+    backendInterface.ctx = summon[Context]
+    val generated = codeGen.genUnit(ctx.compilationUnit)
+    // In Scala 2 backend depening on either global optimizations are enabled
+    // post processing might be delayed to build the call graph, or run imiedietlly if opts are disabled
+    // In Scala 3 we don't perform backend optimizations and though always post process without the delay
+    // https://github.com/scala/scala/pull/6057
```

Comment:
There are a few typos in there and missing punctuation, so it's actually a bit hard to understand the true meaning of this comment.